frontend/catalyst/autograph/ag_primitives.py

```diff
@@ -600,6 +605,131 @@ def set_item(target, i, x):
     return target
 
 
+def update_item_with_add(target, i, x):
```

Comment:
I wonder if it would be more maintainable to have a single function which dispatches to each operator, say in case we wanted to expand the functionality a bit more. What is your opinion on it?

Reply:
I did try to use a helper to generate the different functions, similar to [`_logical_op`](https://github.com/PennyLaneAI/catalyst/blob/6a6fd4f2f915781fb6d425e4d95a92c5a117f31c/frontend/catalyst/autograph/ag_primitives.py#L434), but I couldn't figure out a nice way to put `+=` and friends in a lambda, so this seemed like the next best option.

A possibly more maintainable alternative would be to have a single `update_item_with_op` function that takes an extra parameter and uses that at runtime to determine which JAX function / operator to dispatch to, something like
```python
def update_item_with_op(op, target, i, x):
    if op == '+':
        if isinstance(target, DynamicJaxprTracer):
            target = target.at[i].add(x)
        else:
            target[i] += x
    elif op == '*':
        if isinstance(target, DynamicJaxprTracer):
            target = target.at[i].multiply(x)
        else:
            target[i] *= x
    elif op == '/':
        ...
```
When generating the call in `operator_update.transform` we would just stick in the appropriate constant. I do see two cons to this approach: first, it relies on a set of magic constants, and second, it adds extra compares and branches in the output that could negatively impact performance. Those may not be big concerns though; the calls are only generated from one place and this function isn't exposed to the user (right?) so magic constants aren't that problematic IMO, and if the JAX compiler is smart enough with inlining / constant propagation / dead code elimination it will get rid of the extra compares and branches.

Reply:
Thanks Spencer, that's a good analysis. The single `update_item_with_op` is what I had in mind, but your reasoning is sound so I'm happy to leave the functions as they are :)

> if the JAX compiler is smart enough with inlining / constant propagation / dead code elimination it will get rid of the extra compares and branches

Something that might be interesting to you is that JAX is _not_ capable of compiling most Python code. Instead, it will compile an instruction if either:
- it is function from the jax library (like e.g. `jax.numpy.abs(x)`)
- it is a Python operator acting on a jax data (array) object (like the mult in `jax.numpy.array(2) * 3`)
Consequently, the comparisons and branches all happen before the JAX compiler kicks in (when Python executes the code). There are however [special functions](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.cond.html) that allow you compile a conditional (if statement) for example.